// ng-terminal/esm2015/lib/ng-terminal.module.js
import { NgModule } from '@angular/core';
import { NgTerminalComponent } from './ng-terminal.component';
import { ResizableModule } from 'angular-resizable-element';
import { GlobalStyleComponent } from './global-style/global-style.component';
import { CommonModule } from '@angular/common';
import * as i0 from "@angular/core";
export class NgTerminalModule {
}
NgTerminalModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: NgTerminalModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgTerminalModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: NgTerminalModule, declarations: [NgTerminalComponent, GlobalStyleComponent], imports: [ResizableModule, CommonModule], exports: [NgTerminalComponent] });
NgTerminalModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: NgTerminalModule, imports: [[
            ResizableModule, CommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: NgTerminalModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgTerminalComponent, GlobalStyleComponent],
                    imports: [
                        ResizableModule, CommonModule
                    ],
                    exports: [NgTerminalComponent]
                }]
        }] });
